docs/source/etcd-architecture.rst

```diff
@@ -0,0 +1,238 @@
+they see any changes that have been made, allowing them to respond to those
+changes in a timely manner.
+
+The etcd component is distributed across the entire deployment. It is divided
+into two groups of machines: the core cluster, and the proxies.
+
+In Calico, we deploy an etcd cluster. In small deployments this can be an etcd
+cluster of one, which provides no redundancy but is simple and low-overhead.
+In larger deployments we scale this up, as per the `etcd admin guide`_.
+
+Additionally, on each machine that hosts either a
+:ref:`calico-felix-component` or a :ref:`calico-orchestrator-plugin`, we run
+an etcd proxy. This is an attempt to reduce load on the core cluster and to
+ensure that nodes are shielded from the specifics of the etcd cluster.
+
+etcd is responsible for performing all of the following tasks:
+
+.. _etcd admin guide: https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#optimal-cluster-size
+
+Data Storage
+~~~~~~~~~~~~
+
+etcd stores the data for the Calico network in a distributed, consistent,
+fault-tolerant manner. This set of properties ensures that the Calico network
+is always in a known-good state, while allowing for some number of the machines
+hosting etcd to fail or become unreachable.
+
+This distributed storage of Calico data also improves the ability of the Calico
+components to read from the database (which is their most common operation), as
+they can distribute their reads around the cluster.
+
+Communication
+~~~~~~~~~~~~~
+
+etcd is also used as a communication bus between components through the
+mechanism of having various components watch keys for changes. This allows the
+act of committing state to the database to cause that state to be programmed
+into the network.
+
+
+.. _calico-bgp-component:
+
+BGP Client (BIRD)
+-----------------
+
+Calico deploys a BGP client on every node that also hosts a
+:ref:`calico-felix-component`. The role of the BGP client is to read routing
+state that :ref:`calico-felix-component` programs into the kernel and
+distribute it around the data center.
+
+In Calico, this BGP component is most commonly `BIRD`_, though any BGP client
+that can draw routes from the kernel and distribute them is suitable in this
+role.
+
+The BGP client is responsible for performing the following tasks:
+
+.. _BIRD: http://bird.network.cz/
+
+Route Distribution
+~~~~~~~~~~~~~~~~~~
+
+When :ref:`calico-felix-component` inserts routes into the Linux kernel FIB,
+the BGP client will pick them up and distribute them to the other nodes in the
+deployment. This ensures that traffic is efficiently routed around the
+deployment.
+
+
+.. _calico-bgp-rr-component:
+
+BGP Route Reflector (BIRD)
+--------------------------
+
+For larger deployments, simple BGP can become a limiting factor because it
+requires every BGP client to be connected to every other BGP client in a mesh
+topology. This requires an ever increasing number of connections that rapidly
+become tricky to maintain.
```

Comment:
Well, so does scaling in general...  I think you need to be more precise here, something about an N^2 configuration requirement and N^2 live BGP peering sessions.